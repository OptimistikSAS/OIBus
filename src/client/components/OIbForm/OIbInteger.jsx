import React from 'react'
import PropTypes from 'prop-types'
import { FormGroup, FormFeedback, FormText, Label, Input } from 'reactstrap'

const OIbInteger = ({ label, help, valid, value, name, onChange, defaultValue }) => {
  // if no value was found, load the context with the default value
  // (this will cause a rerender with the correct value)
  React.useLayoutEffect(() => {
    if (value === null) onChange(name, defaultValue)
  }, [value])
  const handleChange = (event) => {
    const { target } = event
    const { value: v } = target
    const newVal = Number.isNaN(parseInt(v, 10)) ? v : parseInt(v, 10)
    onChange(name, newVal, valid(newVal))
  }
  // if no label, we need to minimize the row height
  const style = label ? null : { marginBottom: 0 }
  const validCheck = valid(value)
  return (
    <FormGroup style={style}>
      {label && <Label for={name}>{label}</Label>}
      <Input
        className="oi-form-input"
        type="integer"
        id={name}
        name={name}
        invalid={validCheck !== null}
        value={value || ''}
        onChange={handleChange}
      />
      <FormFeedback>{validCheck}</FormFeedback>
      {help && <FormText>{help}</FormText>}
    </FormGroup>
  )
}
OIbInteger.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  help: PropTypes.element,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  defaultValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  valid: PropTypes.func,
}
OIbInteger.defaultProps = {
  valid: () => null,
  label: null,
  help: null,
  value: null,
}

export default OIbInteger
