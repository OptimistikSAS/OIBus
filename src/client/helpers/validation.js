const type = {
  string: (val) => ((typeof val === 'string' || val instanceof String) ? null : 'value should be a string'),
  number: (val) => ((typeof val === 'number' || val instanceof Number) ? null : 'value should be a number'),
  stringOrNumber: (val) => ((typeof val === 'number' || val instanceof Number)
                        || (typeof val === 'string' || val instanceof String) ? null : 'value should be a number'),
}

const validation = {
  engine: {
    port: (val) => (val >= 1 && val <= 65535 ? null : 'value should be between 1 and 65535'),
    user: (val) => ((val.length > 2) ? null : 'Length should be greater than 2'),
    password: (val) => (/^.{4,}$/.test(val) ? null : 'Length should be greater than 4'),
    filter: (val) => (/^.{2,}$/.test(val) ? null : 'Length should be greated than 2'),
    logParameters: {
      filename: type.string,
      maxsize: type.number,
      maxFiles: (val) => ((val >= 1) && (val <= 10) ? null : 'value should be between 1 and 10'),
      sqliteFilename: type.string,
      sqliteMaxFileSize: type.number,
    },
    scanModes: {
      scanMode: (val) => (/^.{2,}$/.test(val) ? null : 'Length should be greated than 2'),
      cronTime: (val) => (/^.{2,}$/.test(val) ? null : 'Length should be greated than 2'),
    },
    caching: {
      cacheFolder: type.string,
      archiveFolder: type.string,
    },
    proxies: {
      name: (val) => (/^.{2,}$/.test(val) ? null : 'Length should be greater than 2'),
      host: (val) => (/^.{2,}$/.test(val) ? null : 'Length should be greater than 2'),
      port: (val) => (val >= 1 && val <= 65535 ? null : 'value should be between 1 and 65535'),
      username: (val) => (/^.{2,}$/.test(val) ? null : 'Length should be greater than 2'),
      password: (val) => (/^.{2,}$/.test(val) ? null : 'Length should be greater than 2'),
    },
  },
  north: {
    caching: {
      sendInterval: type.number,
      retryInterval: type.number,
      groupCount: type.number,
      maxSendCount: type.number,
    },
    AliveSignal: {
      host: type.string,
      authentication: {
        username: type.string,
        password: type.string,
      },
      id: type.string,
      frequency: type.number,
      proxy: type.string,
    },
    AmazonS3: {
      bucket: type.string,
      folder: type.string,
      authentication: {
        accessKey: type.string,
        secretKey: type.string,
      },
      proxy: type.string,
    },
    InfluxDB: {
      user: type.string,
      password: type.string,
      db: type.string,
      host: type.string,
      precision: type.string,
    },
    OIAnalyticsFile: {
      host: type.string,
      endpoint: type.string,
      authentication: {
        username: type.string,
        password: type.string,
      },
      proxy: type.string,
    },
    OIConnect: {
      host: type.string,
      endpoint: type.string,
      authentication: {
        username: type.string,
        password: type.string,
      },
      proxy: type.string,
    },
    TimescaleDB: {
      user: type.string,
      password: type.string,
      db: type.string,
      host: type.string,
    },
  },
  south: {
    CSV: {
      inputFolder: type.string,
      archiveFolder: type.string,
      errorFolder: type.string,
      separator: type.string,
      timeColumn: type.number,
      points: {
        pointId: type.string,
        value: type.stringOrNumber,
        quality: type.stringOrNumber,
      },
    },
    FolderScanner: {
      inputFolder: type.string,
      minAge: type.number,
      regex: type.string,
    },
    Modbus: {
      host: type.string,
      port: type.number,
      points: {
        pointId: type.string,
        address: type.string,
      },
    },
    MQTT: {
      server: type.string,
      port: type.number,
      username: type.string,
      password: type.string,
      points: {
        pointId: type.string,
        topic: type.string,
      },
    },
    OPCHDA: {
      agentFilename: type.string,
      tcpPort: type.number,
      host: type.string,
      serverName: type.string,
      retryInterval: type.number,
      points: { pointId: type.string },
    },
    OPCUA: {
      host: type.string,
      opcuaPort: type.number,
      httpsPort: type.number,
      endPoint: type.string,
      points: {
        pointId: type.string,
        ns: type.number,
        s: type.string,
      },
    },
    SQLDbToFile: {
      host: type.string,
      post: type.number,
      username: type.string,
      password: type.string,
      database: type.string,
      query: type.string,
      connectionTimeout: type.number,
      requestTimeout: type.number,
      delimiter: type.string,
      filename: type.string,
    },
  },
}

export default validation
