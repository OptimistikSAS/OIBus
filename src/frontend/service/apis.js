const createErrorMessage = (response) => (
  `${response.status}${response.statusText ? ` - ${response.statusText}` : ''}`
)

const getRequest = async (uri) => {
  let text = ''
  try {
    const response = await fetch(uri, { method: 'GET' })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    text = await response.text()
    return JSON.parse(text)
  } catch (error) {
    console.error(`Request error for ${uri}:${error}`, text)
    throw new Error(error)
  }
}

const postRequest = async (uri, body) => {
  try {
    const response = await fetch(uri, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return JSON.parse(await response.text())
  } catch (error) {
    throw new Error(error)
  }
}

const deleteRequest = async (uri) => {
  try {
    const response = await fetch(uri, {
      method: 'DELETE',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return response
  } catch (error) {
    throw new Error(error)
  }
}

const putRequest = async (uri, body) => {
  try {
    const response = await fetch(uri, {
      method: 'PUT',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return response
  } catch (error) {
    throw new Error(error)
  }
}

const getConfig = () => getRequest('/config')
const updateConfig = (body) => putRequest('/config', body)
const activateConfig = () => putRequest('/config/activate')

const createHistoryQuery = async (body) => postRequest('/history-queries', body)
const getHistoryQueries = async () => getRequest('/history-queries')
const getHistoryQueryById = async (id) => getRequest(`/history-queries/${id}`)
const updateHistoryQuery = async (id, body) => putRequest(`/history-queries/${id}`, body)
const enableHistoryQuery = async (id, body) => putRequest(`/history-queries/${id}/enable`, body)
const orderHistoryQuery = async (id, body) => putRequest(`/history-queries/${id}/order`, body)
const deleteHistoryQuery = async (id, position) => deleteRequest(`/history-queries/${id}?position=${position}`)
const getLastCompletedForHistoryQuery = async (id) => getRequest(`/history-queries/${id}/status`)

const getLogs = (fromDate, toDate, pageNumber, pageSize, verbosity) => getRequest(
  `/logs?fromDate=${fromDate || ''}&toDate=${toDate || ''}&pageNumber=${pageNumber}&pageSize=${pageSize}&verbosity=[${verbosity}]`,
)
const getOIBusInfo = () => getRequest('/info')
const getSouthTypes = () => getRequest('/api/installed-south')
const getNorthTypes = () => getRequest('/api/installed-north')
const getSouth = async (id) => getRequest(`/api/south/${id}`)
const getNorth = async (id) => getRequest(`/api/north/${id}`)
const getSouthStatus = (id) => getRequest(`/south/${id}/status`)
const getNorthStatus = (id) => getRequest(`/north/${id}/status`)
const reload = () => getRequest('/reload')
const shutdown = () => getRequest('/shutdown')

export default {
  getSouthTypes,
  getNorthTypes,
  getConfig,
  activateConfig,
  updateConfig,
  getHistoryQueries,
  getHistoryQueryById,
  createHistoryQuery,
  updateHistoryQuery,
  enableHistoryQuery,
  orderHistoryQuery,
  deleteHistoryQuery,
  getLastCompletedForHistoryQuery,
  getLogs,
  getOIBusInfo,
  getSouthStatus,
  getSouth,
  getNorthStatus,
  getNorth,
  reload,
  shutdown,
}
