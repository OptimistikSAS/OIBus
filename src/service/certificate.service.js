import fs from 'node:fs/promises'
import path from 'node:path'

/**
 * Class used to manage certificate files and their content
 */
export default class CertificateService {
  constructor(logger) {
    this.logger = logger
    this.privateKey = null
    this.cert = null
    this.ca = null
  }

  async init(privateKeyFilePath, certFilePath, caFilePath) {
    if (privateKeyFilePath) {
      try {
        this.privateKey = await fs.readFile(path.resolve(privateKeyFilePath))
      } catch (error) {
        this.logger.error(`Key file "${privateKeyFilePath}" does not exist: ${error}`)
      }
    }
    if (certFilePath) {
      try {
        this.cert = await fs.readFile(path.resolve(certFilePath))
      } catch (error) {
        this.logger.error(`Cert file "${certFilePath}" does not exist: ${error}`)
      }
    }
    if (caFilePath) {
      try {
        this.ca = await fs.readFile(path.resolve(caFilePath))
      } catch (error) {
        this.logger.error(`CA file "${caFilePath}" does not exist: ${error}`)
      }
    }
  }
}
