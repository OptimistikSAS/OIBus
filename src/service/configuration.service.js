import path from 'node:path'
import fs from 'node:fs/promises'

import EncryptionService from './encryption.service.js'
import { filesExists } from './utils.js'
import defaultConfig from '../config/default-config.json'

const KEYS_FOLDER = './keys'
const CERTS_FOLDER = './certs'

/**
 * Class responsible for managing the configuration.
 * @class ConfigurationService
 */
export default class ConfigurationService {
  /**
   * Constructor for ConfigService
   * @constructor
   * @param {String} configFilePath - The config file
   * @param {String} cacheFolder - The cache folder to use
   * @return {void}
   */
  constructor(configFilePath, cacheFolder) {
    this.encryptionService = EncryptionService.getInstance()
    this.configFilePath = path.resolve(configFilePath)
    this.cacheFolder = path.resolve(cacheFolder)

    this.keyFolder = path.resolve(KEYS_FOLDER)
    this.certFolder = path.resolve(CERTS_FOLDER)

    this.config = null
    this.modifiedConfig = null
  }

  /**
   * Init the config service by creating folders and initiating default config if necessary
   * @returns {Promise<void>} - The result promise
   */
  async init() {
    let config
    if (await filesExists(this.configFilePath)) {
      config = JSON.parse(await fs.readFile(this.configFilePath, { encoding: 'utf8' }))
    } else {
      config = defaultConfig
    }

    const previousConfig = JSON.stringify(config)
    // When a specific config is used with unencrypted secrets, it allows to encrypt credentials and update the config.
    // In this case, no backup is used to avoid plain text secrets
    // This can happen after a fresh installation with a config file edited by the user or an installation script
    await this.updateConfig(config)

    if (JSON.stringify(this.modifiedConfig) !== previousConfig) {
      await this.activateConfiguration(false)
    } else {
      this.config = this.modifiedConfig
    }
  }

  /**
   * Return the engine, south and north config
   * @returns {Object} - The configs
   */
  getConfig() {
    return {
      engineConfig: this.config.engine,
      southConfig: this.config.south,
      northConfig: this.config.north,
    }
  }

  /**
   * Get active configuration.
   * @returns {Object} - The active configuration
   */
  getActiveConfiguration() {
    return this.config
  }

  /**
   * Update configuration and encrypt password and secrets
   * @param {Object} newConfig - The updated configuration
   * @returns {Promise<void>} - The result promise
   */
  async updateConfig(newConfig) {
    await this.encryptionService.encryptSecrets(newConfig.engine)

    await newConfig.north.reduce((promise, north) => promise.then(
      async () => this.encryptionService.encryptSecrets(north),
    ), Promise.resolve())
    await newConfig.south.reduce((promise, south) => promise.then(
      async () => this.encryptionService.encryptSecrets(south),
    ), Promise.resolve())
    this.modifiedConfig = newConfig
  }

  /**
   * Activate the configuration, i.e. saving a new config file
   * @param {Boolean} backup - Should save a backup file or not
   * @returns {void}
   */
  async activateConfiguration(backup = true) {
    if (backup) {
      const timestamp = new Date().getTime()
      const backupFilename = `${path.parse(this.configFilePath).name}-${timestamp}${path.parse(this.configFilePath).ext}`
      const backupPath = path.resolve(path.parse(this.configFilePath).dir, backupFilename)

      await fs.copyFile(this.configFilePath, backupPath)
    }
    await fs.writeFile(this.configFilePath, JSON.stringify(this.modifiedConfig, null, 4), 'utf8')
    await this.removeOrphanCacheFolders(this.modifiedConfig)
    this.config = JSON.parse(JSON.stringify(this.modifiedConfig))
  }

  async removeOrphanCacheFolders(newConfig) {
    const northIdList = newConfig.north.map((north) => north.id)
    const southIdList = newConfig.south.map((south) => south.id)
    const idList = [...northIdList, ...southIdList]

    const dataStreamFolderPath = path.resolve(this.cacheFolder, 'data-stream')
    const pathExists = await filesExists(dataStreamFolderPath)
    if (pathExists) {
      const folders = await fs.readdir(dataStreamFolderPath)
      await Promise.allSettled(folders.map(async (folder) => {
        const uid = folder.replace('north-', '').replace('south-', '')
        if (!idList.includes(uid)) {
          await fs.rm(path.resolve(dataStreamFolderPath, folder), { recursive: true })
        }
      }))
    }
  }
}
