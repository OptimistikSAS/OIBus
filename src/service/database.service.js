const db = require('better-sqlite3')

const CACHE_TABLE_NAME = 'cache'
const PAGE_SIZE = 50

/**
 * Initiate SQLite database and create the cache table.
 * @param {string} databasePath - The database file path
 * @return {object} - The SQLite database
 */
const createConfigDatabase = (databasePath) => {
  const database = db(databasePath)

  const query = `CREATE TABLE IF NOT EXISTS ${CACHE_TABLE_NAME} (`
                   + 'id INTEGER PRIMARY KEY, '
                   + 'name TEXT UNIQUE, '
                   + 'value TEXT);'
  database.prepare(query).run()

  return database
}
/**
 * Upsert config entry.
 * @param {object} database - The database to use
 * @param {string} name - The config entry
 * @param {string} value - The config value
 * @return {void}
 */
const upsertConfig = (database, name, value) => {
  const query = `INSERT INTO ${CACHE_TABLE_NAME} (name, value) `
                 + 'VALUES (?, ?) '
                 + 'ON CONFLICT(name) DO UPDATE SET value = ?'
  database.prepare(query).run(name, value, value)
}

/**
 * Get configuration.
 * @param {object} database - The database to use
 * @param {string} name - The config name
 * @return {string|null} - The config value
 */
const getConfig = (database, name) => {
  const query = 'SELECT value '
                 + `FROM ${CACHE_TABLE_NAME} `
                 + 'WHERE name = ?'
  const result = database.prepare(query).get(name)

  return result?.value
}

/**
 * Get logs.
 * @param {string} databasePath - The database path
 * @param {string} fromDate - Start date
 * @param {string} toDate - End date
 * @param {string[]} verbosity - Verbosity levels
 * @return {object[]} - The logs
 */
const getLogs = (databasePath, fromDate, toDate, verbosity) => {
  const database = db(databasePath)
  const query = 'SELECT * FROM logs '
                 + 'WHERE timestamp BETWEEN ? AND ? '
                 + `AND level IN (${verbosity.map((_) => '?')})`
  return database.prepare(query).all([fromDate, toDate, ...verbosity])
}

/**
 * Get paginated logs
 * @param {string} databasePath - The database path
 * @param {string} fromDate - Start date
 * @param {string} toDate - End date
 * @param {string} scope - The scope of the log
 * @param {string} textMessage - Text the message must contain
 * @param {string[]} verbosity - Verbosity levels
 * @param {'ASC' | 'DESC'} sorting - The sorting to use
 * @param {number} pageNumber - The page number to request
 * @return {
 * {
 *  content: {timestamp: String, level: String, scope: String, source: String}[],
 *  pageSize: Number,
 *  pageNumber: Number,
 *  totalNumberOfElements: Number,
 *  totalNumberOfPages: Number
 * }} - The paginated logs
 */
const getPaginatedLogs = (
  databasePath,
  fromDate,
  toDate,
  scope,
  textMessage,
  verbosity,
  sorting,
  pageNumber = 0,
) => {
  const database = db(databasePath)
  let whereClause = `WHERE timestamp BETWEEN '${fromDate}' AND '${toDate}' `
  + `AND level IN (${verbosity.map((verb) => `'${verb}'`)})`
  if (scope) {
    whereClause += ` AND scope = '${scope}'`
  }
  if (textMessage) {
    whereClause += ` AND message like '%${textMessage}%'`
  }

  const query = `SELECT timestamp, level, scope, source, message FROM logs ${whereClause}`
      + ` ORDER BY timestamp ${sorting} LIMIT ${PAGE_SIZE} OFFSET ${PAGE_SIZE * pageNumber}`
  const results = database.prepare(query).all()
  const totalNumberOfElements = database.prepare(`SELECT COUNT(*) as count FROM logs ${whereClause}`).get().count
  const totalNumberOfPages = Math.ceil(totalNumberOfElements / PAGE_SIZE)

  return {
    content: results,
    pageSize: results.length,
    pageNumber,
    totalNumberOfElements,
    totalNumberOfPages,
  }
}

/**
 * Get South connector related data for HistoryQuery.
 * @param {string} databasePath - The database path
 * @returns {object} - The data
 */
const getHistoryQuerySouthData = (databasePath) => {
  const database = db(databasePath)
  const query = `SELECT * FROM ${CACHE_TABLE_NAME}`
  return database.prepare(query).all()
}

module.exports = {
  createConfigDatabase,
  upsertConfig,
  getConfig,
  getLogs,
  getPaginatedLogs,
  getHistoryQuerySouthData,
}
