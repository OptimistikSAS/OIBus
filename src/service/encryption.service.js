import os from 'node:os'
import crypto from 'node:crypto'
import fs from 'node:fs/promises'
import path from 'node:path'

import selfSigned from 'selfsigned'

import { createFolder, filesExists } from './utils.js'

const CERT_PRIVATE_KEY_FILE_NAME = 'privateKey.pem'
const CERT_PUBLIC_KEY_FILE_NAME = 'publicKey.pem'
const CERT_FILE_NAME = 'cert.pem'

const OIBUS_PRIVATE_KEY_FILE_NAME = 'private.pem'
const OIBUS_PUBLIC_KEY_FILE_NAME = 'public.pem'

/**
 * Service used to manage encryption and decryption of secrets in the config file
 * Also responsible to create private and public key used for encrypting the secrets
 */
export default class EncryptionService {
  static getInstance() {
    if (!EncryptionService.instance) {
      EncryptionService.instance = new EncryptionService()
    }
    return EncryptionService.instance
  }

  constructor() {
    this.keyFolder = null
    this.certsFolder = null
  }

  /**
   * Set the key folder path.
   * @param {String} keyFolder - The folder path to store the keys
   * @returns {void}
   */
  setKeyFolder(keyFolder) {
    this.keyFolder = keyFolder
  }

  /**
   * Set the cert folder path.
   * @param {String} certsFolder - The folder path to store the keys
   * @returns {void}
   */
  setCertsFolder(certsFolder) {
    this.certsFolder = certsFolder
  }

  /**
   * Check if local certificates exist and create them if not.
   * @returns {Promise<void>} - The result promise
   */
  async checkOrCreateCertFiles() {
    const privateKeyPath = path.resolve(this.certsFolder, CERT_PRIVATE_KEY_FILE_NAME)
    const publicKeyPath = path.resolve(this.certsFolder, CERT_PUBLIC_KEY_FILE_NAME)
    const certPath = path.resolve(this.certsFolder, CERT_FILE_NAME)
    await createFolder(this.certsFolder)

    if (!await filesExists(privateKeyPath) || !await filesExists(publicKeyPath) || !await filesExists(certPath)) {
      const certificate = selfSigned.generate([
        { name: 'commonName', value: 'OIBus' },
        { name: 'countryName', value: 'FR' },
        { name: 'stateOrProvinceName', value: 'Savoie' },
        { name: 'localityName', value: 'Chambery' },
        { name: 'organizationName', value: 'Optimistik' },
        { name: 'organizationalUnitName', value: 'R&D' },
      ], {
        keySize: 2048,
        days: 36500,
        algorithm: 'sha256',
        pkcs7: true,
        extensions: [
          {
            name: 'basicConstraints',
            cA: false,
          },
          {
            name: 'keyUsage',
            keyCertSign: true,
            digitalSignature: true,
            nonRepudiation: true,
            keyEncipherment: true,
            dataEncipherment: true,
          },
          {
            name: 'extKeyUsage',
            clientAuth: true,
            serverAuth: true,
          },
          {
            name: 'subjectAltName',
            altNames: [
              {
                type: 6, // URI
                value: `urn:${os.hostname()}:OIBus`,
              },
              {
                type: 2, // DNS
                value: os.hostname(),
              },
            ],
          },
        ],
      })

      await fs.writeFile(privateKeyPath, certificate.private)
      await fs.writeFile(publicKeyPath, certificate.public)
      await fs.writeFile(certPath, certificate.cert)
    }
  }

  /**
   * Check if private/public keys exist and create them if not.
   * @returns {Promise<void>} - The result promise
   */
  async checkOrCreatePrivateKey() {
    const privateKeyPath = path.resolve(this.keyFolder, OIBUS_PRIVATE_KEY_FILE_NAME)
    const publicKeyPath = path.resolve(this.keyFolder, OIBUS_PUBLIC_KEY_FILE_NAME)

    await createFolder(this.keyFolder)

    if (!await filesExists(privateKeyPath) || !await filesExists(publicKeyPath)) {
      const {
        privateKey,
        publicKey,
      } = crypto.generateKeyPairSync('rsa', {
        modulusLength: 4096,
        publicKeyEncoding: {
          type: 'spki',
          format: 'pem',
        },
        privateKeyEncoding: {
          type: 'pkcs8',
          format: 'pem',
          cipher: 'aes-256-cbc',
          passphrase: '',
        },
      })

      await fs.writeFile(privateKeyPath, privateKey)
      await fs.writeFile(publicKeyPath, publicKey)
    }
  }

  /**
   * Recursively iterate through an object tree and encrypt sensitive fields.
   * @param {object} configEntry - The object to iterate through
   * @returns {Promise<void>} - The result promise
   */
  async encryptSecrets(configEntry) {
    if (configEntry) {
      await Object.entries(configEntry).reduce((promise, [key, value]) => promise.then(
        async () => {
          if (typeof value === 'object') {
            await this.encryptSecrets(value)
          } else if (['password', 'secretKey', 'token', 'secret', 'sasToken', 'accessKey', 'clientSecret'].includes(key)
              && value.startsWith('{{notEncrypted}}')) {
            configEntry[key] = await this.encryptText(value.replace('{{notEncrypted}}', ''))
          }
        },
      ), Promise.resolve())
    }
  }

  /**
   * Return the encrypted text
   * @param {String} plainText - The text to encrypt
   * @return {Promise<String>} - The encrypted text
   */
  async encryptText(plainText) {
    const absolutePath = path.resolve(this.keyFolder, OIBUS_PUBLIC_KEY_FILE_NAME)
    const publicKey = await fs.readFile(absolutePath, 'utf8')
    const buffer = Buffer.from(plainText, 'utf8')
    const encrypted = crypto.publicEncrypt(publicKey, buffer)
    return encrypted.toString('base64')
  }

  /**
   * Return the decrypted text
   * @param {String} encryptedText - The text to decrypt
   * @return {Promise<String>} - The decrypted text
   */
  async decryptText(encryptedText) {
    const absolutePath = path.resolve(this.keyFolder, OIBUS_PRIVATE_KEY_FILE_NAME)
    const privateKey = await fs.readFile(absolutePath, 'utf8')
    const buffer = Buffer.from(encryptedText, 'base64')
    const decrypted = crypto.privateDecrypt(
      {
        key: privateKey.toString(),
        passphrase: '',
        padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
      },
      buffer,
    )
    return decrypted.toString('utf8')
  }
}
