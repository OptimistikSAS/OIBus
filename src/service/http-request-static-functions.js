import url from 'node:url'
import { createReadStream } from 'node:fs'
import path from 'node:path'

import FormData from 'form-data'
import ProxyAgent from 'proxy-agent'
import fetch from 'node-fetch'
import https from 'node:https'

/**
 * Create a proxy agent to use wih HTTP requests
 * @param {'http' | 'https'} protocol - The protocol to use
 * @param {String} host - The host
 * @param {Number} port - The port
 * @param {String} username - The username
 * @param {String} password - The decrypted password
 * @param {Boolean} acceptUnauthorized - Reject unauthorized certificate
 * @returns {ProxyAgent.ProxyAgent} - The ProxyAgent to use
 */
const createProxyAgent = (protocol, host, port, username, password, acceptUnauthorized = false) => {
  const proxyOptions = url.parse(`${protocol}://${host}:${port}`)

  if (username && password) {
    proxyOptions.auth = `${username}:${password}`
  }
  proxyOptions.rejectUnauthorized = !acceptUnauthorized

  return new ProxyAgent(proxyOptions)
}

/**
 * Mutate the headers to add authentication headers according to the type of authentication
 * @param {Object} headers - The headers to save the authentication
 * @param {'Basic' | 'API Key' | 'Bearer'} type - The type of authentication
 * @param {String} key - The key
 * @param {String} secret - The decrypted secret
 * @return {void}
 */
const addAuthenticationToHeaders = (headers, type, key, secret) => {
  switch (type) {
    case 'Basic':
      headers.Authorization = `Basic ${Buffer.from(`${key}:${secret}`).toString('base64')}`
      break

    case 'API Key':
      headers[key] = secret
      break

    case 'Bearer':
      headers.Authorization = `Bearer ${secret}`
      break

    default:
      throw new Error(`Unrecognized authentication type: "${type}".`)
  }
}

/**
 * Send the request using node-fetch
 * If "headers" contains Content-Type "data" is sent as string in the body.
 * If "headers" doesn't contain Content-Type "data" is interpreted as a path and sent as a file.
 * @param {String} requestUrl - The URL to send the request to
 * @param {String} method - The request type
 * @param {Object} headers - Headers to send
 * @param {String} data - The body or file to send
 * @param {Number} timeout - The timeout in seconds
 * @param {Object} proxyAgent - Proxy to use
 * @param {Boolean} acceptUnauthorized - Reject unauthorized certificate
 * @return {Promise<void>} - The result promise
 */
const httpSend = async (
  requestUrl,
  method,
  headers,
  data,
  timeout,
  proxyAgent,
  acceptUnauthorized = false,
) => {
  let body
  let readStream
  if (Object.prototype.hasOwnProperty.call(headers, 'Content-Type')) {
    body = data
  } else {
    readStream = createReadStream(data)
    body = new FormData()

    // Remove timestamp from the file path
    const { name, ext } = path.parse(data)
    const filename = name.slice(0, name.lastIndexOf('-'))

    const bodyOptions = { filename: `${filename}${ext}` }
    body.append('file', readStream, bodyOptions)

    const formHeaders = body.getHeaders()
    Object.keys(formHeaders).forEach((key) => {
      headers[key] = formHeaders[key]
    })
  }

  const fetchOptions = {
    method,
    headers,
    body,
    timeout: timeout * 1000,
  }
  if (proxyAgent) {
    fetchOptions.agent = proxyAgent
  } else if (acceptUnauthorized && requestUrl.startsWith('https')) {
    fetchOptions.agent = new https.Agent({ rejectUnauthorized: false })
  }

  let response
  try {
    response = await fetch(requestUrl, fetchOptions)
  } catch (error) {
    readStream?.close()
    const requestError = error
    requestError.responseError = false
    throw requestError
  }
  readStream?.close()
  if (!response.ok) {
    const responseError = new Error(response.statusText)
    responseError.responseError = true
    responseError.statusCode = response.status
    throw responseError
  }
}

export {
  createProxyAgent,
  addAuthenticationToHeaders,
  httpSend,
}
