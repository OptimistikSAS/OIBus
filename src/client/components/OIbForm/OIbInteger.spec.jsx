import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import OIbInteger from './OIbInteger.jsx'
import { minValue } from '../../../services/validation.service'

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('OIbInteger', () => {
  test('check OIbInteger with value="a text"', () => {
    act(() => {
      ReactDOM.render(<OIbInteger
        label="label"
        value="12345678"
        name="name"
        onChange={() => (1)}
        defaultValue="0"
      />, container)
    })
    expect(container).toMatchSnapshot()
  })
  test('check Texte with too small value', () => {
    act(() => {
      ReactDOM.render(<OIbInteger
        label="label"
        name="name"
        onChange={() => (1)}
        value="12"
        valid={minValue(100)}
        defaultValue="0"
      />, container)
    })
    expect(container).toMatchSnapshot()
  })
})
