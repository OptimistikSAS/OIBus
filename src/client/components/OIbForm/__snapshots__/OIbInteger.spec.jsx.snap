// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OIbInteger check OIbInteger with value="a text" 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      class=""
      for="name"
    >
      label
    </label>
    <input
      class="oi-form-input form-control"
      id="name"
      name="name"
      type="integer"
      value="12345678"
    />
    <div
      class="invalid-feedback"
    />
  </div>
</div>
`;

exports[`OIbInteger check Texte with too small value 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      class=""
      for="name"
    >
      label
    </label>
    <input
      class="oi-form-input is-invalid form-control"
      id="name"
      name="name"
      type="integer"
      value="12"
    />
    <div
      class="invalid-feedback"
    >
      Value should not greater than 100
    </div>
  </div>
</div>
`;
